File: ts-tests/src/pages/login-page.ts
import { Page, Locator, expect } from '@playwright/test';

export class LoginPage {
  readonly page: Page;
  readonly usernameInput: Locator;
  readonly passwordInput: Locator;
  readonly loginButton: Locator;
  readonly successHeading: Locator;

  constructor(page: Page) {
    this.page = page;
    this.usernameInput = page.getByLabel('Username');
    this.passwordInput = page.getByLabel('Password');
    this.loginButton = page.getByRole('button', { name: /login/i });
    this.successHeading = page.getByRole('heading', {
      name: 'You logged into a secure area!',
    });
  }

  async open() {
    await this.page.goto('/login');
  }

  async login(username: string, password: string) {
    await this.usernameInput.fill(username);
    await this.passwordInput.fill(password);
    await this.loginButton.click();
  }

  async isLoggedIn(): Promise<boolean> {
    // Mirror Java heuristic: URL contains /secure or heading visible
    const url = this.page.url();
    if (url.includes('/secure')) {
      return true;
    }
    try {
      return await this.successHeading.isVisible();
    } catch {
      return false;
    }
  }
}

File: py-tests/pages/login_page.py
from playwright.sync_api import Page, Locator


class LoginPage:
    def __init__(self, page: Page) -> None:
        self.page = page
        self.username_input: Locator = page.get_by_label("Username")
        self.password_input: Locator = page.get_by_label("Password")
        self.login_button: Locator = page.get_by_role("button", name="Login")
        self.success_heading: Locator = page.get_by_role(
            "heading", name="You logged into a secure area!"
        )

    def open(self) -> "LoginPage":
        self.page.goto("https://practice.expandtesting.com/login")
        return self

    def login(self, username: str, password: str) -> "LoginPage":
        self.username_input.fill(username)
        self.password_input.fill(password)
        self.login_button.click()
        return self

    def is_logged_in(self) -> bool:
        url = self.page.url
        if "/secure" in url:
            return True
        try:
            return self.success_heading.is_visible()
        except Exception:
            return False
